# Century From Year 


Given a year, return the century it is in. The first century spans from the year 1 up to and including the year 100, the second- from the year 101 up to and including the year 200, etc. 


* Example year = 1905, the output be: 
    centuryFromYear(year) = 20

* For year 1700, the output should be:
    centuryFromYear(year) = 17




In [1]:
# Think of every year as 100 

1900/100 

19.0

In [2]:
1905/100

19.05

In [3]:
2000/100

20.0

In [4]:
2020/100

20.2

We see that it is 20.2, but in reality it should be 21, because there is no such thing as .2 of a year. If there is a decimal you need to round it UP to the nearest #. 

In [10]:
import math 
math.ceil(2020/100)

21

In [12]:
print(math.ceil(2020/100))

21


As metioned earlier we need to use ``ceil`` in order to deomonstrate that there is not a .2 of a year. We need to round it up if there is a decimal in the year. In 2022 we are currently in the 21st century. That is how we know that the progrm works.

In [13]:
math.ceil(1900/100)

19

In [14]:
math.ceil(1905/100)


20

Now that we have the understanding of how the function ``centuryFromYear()`` should work we now need to 'code' it 

In [17]:
def centuryFromYear(year):
    return math.ceil(1905 / 100)




## Outcome: 
The solution above works, you don't need to worry about the importing because in the problem set they already have it done for you. 

# Check Palindrome 

There are multiple ways to check if a word is a palidrome take the following: 

$$ aabbaa $$

We can first start by putting a pointer on the first letter and the second letter. 

$$
[a]aba[a]
$$

then we can check if those two characters match. They do, so we continue with the next characters 

$$
a[a]b[a]a
$$

Because both of the cases do come out as true, we can stop there because that means that it is a palindrome with that word. No mater how many times we flip the word the letters will stay the same. 

Now that we understand the logic we are to code it under the function ``checkPalindrome()``:


In [18]:
def checkPalindrome(inputString):
    return inputString[::-1] == inputString

Above we see that by using the ``inputString[::-1]`` we basically flip the word that is stored in the inputString variable. This is one of the approach. There is another approach that can be taken and that will be explained below: 

In [20]:
def checkPalindrome(inputString):
    i = 0 
    j = len(inputString) - 1
    while i < j: 
        if inputString[i] != inputString[j]: return False
        i += 1 
        j -= 1
    return True 

### Outcome: 

This approach is similar to the logic that we used at the start of the problem, where we use pointers in order ot see if letters match. We stop when we have gone through the lenght of the word. The way that we got that is by using the ``len(inputString) - 1``. At the end if we manage to get through the whole length of the word and the characters match then, it will return true; therefore, a palindrome 

# Adjacent Elements Product 

Givena n array of integers, find the pair of adjacent elements that has the largest product and return that product. 

* Example:
For ``inputArray = [3,6,-2,-5,7,3]``, the output should be ``adjacentElementsPorduct(inputArray) = 21 ``

``7`` and ``3`` produce the largest product 

In [24]:
def adjacentElementsProduct(inputArray):
    """
    We are mapping the array and finding the two bigest #s
    """
    return max(
        [inputArray[i] * inputArray[i + 1] for i in range(0,len(inputArray) -1)]
    )

For the ``inputArray[i] * inputArray[i +1]`` is where we are going to be selecting the numbers, we take the first number of the ``inputArray`` and then we take the number after it by doing ``inputArray[i+1]``


And we do this until we reach the limit of the array, and we do so my doing ``range(0,len(inputArray) -1)``. Where we start at position zero, and then we have the end be based on the count of the actual #s in the array. 

# Make Array Consectuive 2

Ratior got status of idfferent sizes as a present from CodeMaster for his birthday, for each statue having an noonegative integer size. Since he likes to make things perfect, he wants to arrange them from samllest to largest so that each stattue will be bigger than the previous one exactly by 1. He may need additional statues to do that, help him figure out the minimm number of additional statues needed. 


* Example: For ``statues = [6,2,3,5]`` the output should be ``makeArrayConsecutive2(statues) = 3``

Ratiorg needs statues of sizes ``4,5,7``





Take the following array: 

$$
    [6,2,3,8]
$$
We want to first re-arrange it: 
$$
    [2,3,6,8]
$$

We now know which numbers are missing in order to have the array go up by 1. 
$$
    4,5,7
$$
We see that the numbers that we need in order to have the array to go up by one is 3. So the finally we should get: 
$$
    [4,5,6,7,8]
$$

### Now to craft the solution: 

In [25]:
def makeArrayConsecutive2(statues):
    s = 0 
    statues = statues.sort() # this is what we spoke about, having the numbers be re-arrange in order to see what #s are missing
    
    for i in range(0, len(statues) -1):
        """
            Here we traverse through the array 
            then we run the if func. where if the the subtraction of the first number and second number
            is greater than one, the difference gets stored into the variabe s, where s represents 
            the ammount of numbers that are needed additonaly in order to have the array go up by one. 
        """
        if statues[i + 1] - statues[i]>1:
            s += statues[i+1]- statues[i] - 1
    return s 


# Matrix Elements Sum 

After beomcing famous, the codeBots decided to move onto a new building together, each room has its own cost, and some are free, but there is a rumor that all of the free rooms are haundted. Since the CodeBots are quite supersitious, they refuse to stay in any other of the free rooms, or any of the rooms that are below the free rooms. 


Give ``matrix``, a rectangular matrix of integers, whereeach value represents the cost of the room, your task is to return the total sum of all the roos that are not suitable for the codebots. (i.e. Add up all the values that don't appear as a 0)

$$

$$

### The logic: 

Based on the matrix provided if there is a ``0`` in the Column then you completely ignore the numbers that are below of the 0. After doing so you can just add all the numbers that remain in the matrix. 

Then we are going to work with finding how many rows we got in 2d dimensional array 

In [27]:
def matrixElementsSum(matrix):
    rows_len = len(matrix)
    cols_len = len(matrix)
    cols = []
    for c in range(cols_len):
        # we are traversing throug the matrix. 
        col = []
        for r in range(rows_len):
            col.append(matrix[r][c])
        cols.append(col)

    for arr in cols: 
        for n in arr: 
            if n == 0: break # meaning we no longer continue counting. 
            s += n 
    return s 

# All Longest Strings 

Given an array of strings, return another array containing all of its longest strings. 

* Example: 

For ``inputArray = ["aba", "aa", "ad", "vcd", "aba"]``, the output should be: 

``allongestStrings(inputArray) = ["aba","vcd","aba"]``

In [2]:
def allLongestStrings(inputArray):
    m = max(inputArray, key= len) #this will store the longest string
    return [s for s in inputArray if len(s) == m]
    # this returns the array with the longest strings. 

A shorter way to do this:

In [3]:
def allLongestStrings(inputArray):
    return [s for s in inputArray if len(s) == len(max(inputArray, key = len))]

# Common Character Count 

Given two strings, find the number of common characters between them 

* Example: for ``s1 = "aabcc" `` and ``s2 ="adcaa"``, the output should be: 

``commonCharacterCount(s1,s2) = 3``

Strings have ``3`` common characters = 2 ``a`` and 1 ``c``

In [8]:
from collections import Counter


def commonCharacterCount(s1,s2):
    """
    We first want to see the len of each of the strings 
    so we are going to make a dictionary to keep count of 
    the ammount of times that there are duplciates, but we can 
    also relate to functions from collections import Counter
    """
    c1 = Counter(s1) # we can iterate strings, returns dictionary 
    c2 = Counter(s2) 
    sum_ = 0 
    seen = set()
    for c in s1: 
        if c in c2 and c not in seen: 
            sum_ += c2[c] if c1[c] > c2[c] else c1[c]
            seen.add(c)
    return sum_ 
            


# isLucky 

Ticket numbers usualy consist of an even number of digits. A ticket number is considered lucky if the sum of the first half of the digits is equal to the sum of the second half. 

Given a ticket number n, determine if it's lucky or not. 

In [11]:
def isLucky(n):
    s = str(n)
    f_half = sum([int(x) for x in list(s[0:int(len(s)) / 2])])
    s_half = sum([int(x) for x in list(s[int(len(s) / 2): len(s)])])
    return f_half == s_half

keep in mind to make sure to read the code from inside out.

# Sort By Height 

Some people are standign in a row in a park. There are trees between them which cannot be moved. Your task is to rearrange the people by their heights in a non descending order without moving the tress. People can be very tall. 

* Example: For ``a[-1,150,190,170,-1,-1,160,180]``, the output should be: 

``sortByHeight(a) = [-1,150,160,170,-1,-1,180, -190]`` 

In [13]:
def sortByHeight(a):
    holder = sorted([x for x in a if x  != -1])
    for i in range(len(a)): 
        """
        if we dont find a number that is not -1 then we replace 
        it with the smallest number 
        """
        if a[i] != -1: 
            a[i] = holder[j]
            j+=1
    return a 


# firstDuplicate

Given an array ``a`` that contians only numbers in the range from ``1`` to ``a.length``, find the first duplicate number for whciht he second occurance has the minimal index. In other words, if there ar emore than 1 duplicated numbers, return the number for whcih the second occurence has a smaller index than the second occurance of the other number does. If there are no such elements, return -1. 

* For example:

For ``a= [2,1,3,5,3,2]``, the output should be ``solution(a)= 3 `` 

there are ``2 `` duplciates: the numbers ``2`` and ``3``. The second occurance of ``3`` has a smaller index than the second occurrence of ``2 `` does, so the answer is ``3``. 

* for ``a= [2,2]``, the output should be ``solution(a) = 2 ``

* For ``a= [2,4,3,5,1]``, the output should be solution(a) = -1

In [ ]:
def solution(a): 
    item= set() # set deosn't allow duplicates
    for num in a: 
        if num in item: 
            return num
        item.add(num)
        
    return -1 # if there are no duplicates. 

# FirstNotRepeatingCharacter 

Given a string ``s`` consisting of small English letters, find and return the first instance of a non- repeating character in it. If there is no such character, return ``'_'``. 

* Example: 
* for ``s ="abacabad"``, the output should be ``solution(s) = 'c'``. 

There are ``2`` non-repeating characters in the string: ``'c'`` and ``'d'``. Return ``'c'`` since it appears in the string first. 

* For ``s = "abacabaabacaba"``, the output should be ``solution(s) = '_'``. 

therea re no characters in this string that do not repeat. 

The solution below was marked as only half correct. 

In [4]:
def nonrepeat(string): 
    freq= {}
    for x in string: 
        """
        What this does is that if the key value exists 
        or if it is new, then it will return x and add +1 
        to the x. Then if we happen to loop to a same 
        character in the string then it will update its x 
        by +1. 
        """
        freq[x] = freq.get(x,0) + 1
    for i in string: 
        if freq[i] == 1: 
            """
            The very frequency character that has 1 
            as its frequency is the first non-repeating 
            character; thus, our answer. 
            """
            return i 
    return -1 

print(nonrepeat('abbcd'))
print(nonrepeat('abbcda'))

a
c


# Rotate Image 

You are given an n x n 2D matrix that represents an imge. Rotate the image by 90 degrees. clockwise. stay under O($n^2$) O(1)

In [7]:
def solution(a):
    a.reverse()
    for i in range(len(a)):
        for j in range(i):
            a[i][j], a[j][i] = a[j][i], a[i][j]
    return a


That is one of the solutions 

In [8]:
def solution(a):
    return list(zip(*reversed(a)))

# traverseTree

Givena  binary tree of integers t, return its node values in the following format: 

* The first eleemtn should be the value of the tree rot; 

* the next element should be the values of th enodes at height 1 (i.e the root children) ordered from the lefrtmost to the rightmost one; 

* the elements after that shoul be the vlaues of the nodes at height 2 (i.e the children of the nodes at height 1) ordered in the same way. 


In [9]:
#
# Binary trees are already defined with this interface:
# class Tree(object):
#   def __init__(self, x):
#     self.value = x
#     self.left = None
#     self.right = None
def solution(t):
    if t is None: 
        return []
    res = [] 
    q= [t]
    while q: 
        v= q.pop(0)
        res += [v.value]
        if v.left: 
            q += [v.left]
        if v.right: 
            q += [v.right]
    return res



In [10]:
# here is another attempt to the problem: 

def solution(t):
    if t is None: 
        return []
    stack = []
    result = []
    stack.append(t)
    while len(stack) > 0: 
        new_tree = stack.pop(0)
        if new_tree.value is not None: 
            result.append(new_tree.value)
        if new_tree.left is not None: 
            stack.append(new_tree.left)
        if new_tree.right is not None: 
            stack.append(new_tree.right)
    return result

# LargestValuesInTreeRows 

You have a binary tree t, your task is to find the largest value in each row of this tree. In a tree, a row is a set of nodes that have equal depth for example, a row iwht depth 0 is a tree root, a row with depth 1 is composed of the root's children, etc. 

Return an array in whciht he first elemetns is the largest value in the row with depth 0, the second element is the largest value in the row with depth 1, the third element is the largest element in the row with depth 2, etc. 

In [11]:
import math 

def solution(t): 
    if t is None: 
        return []
    
    slack = [t]
    result = []
    while len(stack)>0: 
        result.append(max(tree.value for tree in stack))
        next_row= [tree.left for tree in stack if tree.left] + [tree.right for tree in stack if tree.right]
        stack= next_row
    return result
    

Another attempt to this problem 

In [12]:

maxes = []

def solution(t):
    traverseAndLogTree(t, 1)
    return maxes


def traverseAndLogTree(node, row):
    if not node:
        return
    
    if len(maxes) < row:
        maxes.append(node.value)
    
    if maxes[row - 1] < node.value:
        maxes[row - 1] = node.value
    
    traverseAndLogTree(node.left, row + 1)
    traverseAndLogTree(node.right, row + 1)

    

# RemoveKFromList

Given a singly linked list of integer l and an integer k, remove all elements from list l htat have a value equal to k. 

* For ``l = [3,1,2,3,4,5]`` and ``k=3`` the output should be: 

``solution(l,k) = [1,2,4,5];``

* For ``l = [1,2,3,4,5,6,7]`` and ``k= 10``, the output should be: 

``solution(l,k) = [1,2,3,4,5,6,7]``

### Attempt on question

So basically if the k value is a number in the linked list then you need to remove that number that is in the linked list. 

In [7]:
def solution(l, k):
    c = l
    while c:
        if c.next and c.next.value == k:
            c.next = c.next.next
        else:
            c = c.next
    return l.next if l and l.value == k else l

Another attempt on the solution

In [5]:
def solution(l,k):
    curr = l 
    while curr: 
        if curr.next and curr.next.value == k: 
            curr.next= curr.next.next 
        else: 
            curr = curr.next 
    return l if (l is None or l.value != k) else l.next

Another attempt on the solution: 

In [6]:
import sys 
sys.setrecursionlimit(100000)

def solution(l,k):
    if l is None: return 
    # handle first element 
    if l.value == k: 
        l - uniqueNext(l)
    # handle the rest 
    if l is None: return 
    l.next = solution(l.next,k)
    return l 

def uniqueNext(l):
    if l.next is None: return 
    if l.value == l.next.value: 
        return uniqueNext(l.next)
    else: 
        return l.next


# isListPalindrome 

Given a single linked list of integers, determine whether or not it's a palindrome. 

* Note: in examples below and test preview linked list are presented as arrays just for simplicity of visualization: in real data you will be given a head l of linked list

* For ``l = [0,1,0]``, the output should be ``solution(l) =true ``; 

* For ``l = [1,2,2,3]``, the output should be ``solution(l) = false `` 

In [8]:
"""
Definition for singly- linked list:
class ListNode(object):
    def __init__(self, x):
        self.value = x 
        self.next = None
"""

def solution(l):
    if not l or not l.next: 
        return True 
    s= 1 
    n= 1 
    while n.next: 
        n= n.next
        s += 1 

    middle = s//2

    n = l 
    for i in range(middle):
        n = n.next 

    if s % 2: 
        n = n.next 

    r = n # reverse n 
    m = r.next 
    for _ in range(middle-1): # flip n 
        m.next, r, m = r, m, m.next
    
    for _ in range(middle):
        if r.value != l.value: 
            return False 
        r = r.next
        l = l.next   
    return True 

Another Attempt 

In [9]:
def solution(l):
    a = []
    while l != None: 
        a.append(l.value)
        l = l.next 
    return a== a[::-1]

Attempt #2 

In [10]:
def solution(l):
    s= 0 
    k = l 

    while l: 
        l = l.next 
        s += 1 

    l = k 
    i=0 
    p= None 

    while l: 
        i += 1 
        if i> s //2: 
            n = 1 
            l = l.next 
            n.next = p 
            p = n 
        else: 
            l = l.next 
        
    while k and p: 
        if k.value != p.value: 
            return False 
        k = k.next 
        p = p.next 
    return True

# AddTwoHugeNumbers 

You're given ``2`` huge integers represented by linked list. Each linked list element is a number from ``0`` to ``9999`` that represent a number might have leading zeros. Your task is to add up these huge integers and return the result in the same format. 

Example: 
* For ``a = [9876, 5432, 1999]`` and ``b=[1,8001]``, the output shoudl be ``solution(a,b) = [9876,5434,0 ]`` 
Explanation: ``


In [1]:
def solution(a,b):
    a = reverse(a)
    b = reverse(b)

    carry = 0 
    result = None 

    while a is not None or b is not None or carry > 0: 
        raw= ((a.value if a is not None else 0)+ (b.value if b is not None else 0) + carry)
        node = ListNode(raw % 10000)
        node.next = result 
        result = node 
        carry = raw // 10000

        if a is not None: 
            a = a.next 
        if b is not None: 
            b = b.next 
        
        return result 
    
def reverse(list):

    current = list 
    previous = None 

    while current is not None: 
        previous, current.next, current = current, previous
        current, current.next 

    return previous 
    

Another attempt to the problem: 

In [4]:
# Definition for singly- linked list:
# class ListNode(object):
#   def __init__(self, x):
#       self.value = x
#       self.next = None


def solution(a,b):
    str_a = ''
    str_b = ''
    ret = [] 
    while a: 
        str_a = str_a + str(a.value).zfill(4)
        a = a.next 
    while b: 
        str_b = str_b + str(b.value).zfill(4)
        b = b.next 
    ans = str(int(str_a) + int(str_b))
    while ans: 
        ret.insert(0, int(ans[-4:]))
        ans = ans[:-4]
    return ret 

Another Attempt 

In [5]:
def solution(a,b): 
    ar1= []
    ar2=[]
    while a: 
        ar1.append(a.value)
        a = a.next 
    while b: 
        ar2.append(b.value)
        b= b.next 
    ar1= ar1[::-1]
    ar2= ar2[::-1]
    carry = 0 
    res = []
    for i in range(max(len(ar1), len(ar2))):
        if i >= len(ar1) and i< len(ar2):
            x = carry + ar2[i]
        elif i >= len(ar2) and i< len(ar1):
            x= carry +ar1[i]
        elif i < len(ar1) and i< len(ar2):
            x = carry + ar1[i] + ar[i]
        carry = x//10000
        x = x%10000
        res.append(x)
    if carry > 0: 
        res.append(carry)
    return res[::-1]

In [6]:
def solution(a,b):
    aa= ''
    bb= ''
    while a: 
        aa+= '0' *(4-len(str(a.value))) + str(a.value)
        a = a.next 

    while b: 
        bb+= '0' * (4-len(str(b.value))) + str(b.value)
        b = b.next 
    suma = str(int(aa) + int(bb))
    c= '0' *(0 if len(suma)%4 == 0 else 4 -len(sum)%4) +suma
    cc = [int(c[4*i:4*i+4]) for i in range(len(c)//4)]
    node = ListNode(cc[0])
    l = node 
    for i in range(len(cc) -1):
        new_node= ListNode(cc[i+1])
        l.next= new_node
        l = l.next
    return node

# MergeTwoLinkedList 


Givent two singly linked lists sorted in non-decreasing order, your task is to merge them. In other words, return a singly linked list, also sorted in nondecreasing order, that contain the elements from both original lists. 

## Example: 
* For ``l1=[1,2,3]`` and ``l2= [4,5,6]`` the output should be: 
``solution(l1,l2) =[1,2,3,4,5,6]`` 

* For ``l1= [1,1,2,4]`` and ``l2= [0,3,5]`` the outut should be: 
``solution(l1,l2) = [0,1,1,2,3,4,5]``

In [7]:
def solution(l1,l2):
    Headt = ListNode(None)
    current = Headt 
    while l1 and l2: 
        if l1.value <= l2.value:
            current.next =l1
            l1= l1.next 
        else: 
            current.next = l2 
            l2 = l2.next 
        current = current.next 

    if l1:
        current.next= l1 
    elif l2: 
        current.next = l2 
    
    return Headt.next

Another Example: 

In [8]:
"""
Definition for singly- linked list: 


class ListNode(object):
    def __init__(self, x):
        self.value = x 
        self.next = None
"""

def solution(l1,l2):
    dummy = ListNode('Head')
    x = dummy 
    while l1 != None or l2 != None: 
        if l1 != None and l2 != None: 
            val1 = l1.value 
            val2 = l2.value 
            if val1 <= val2: 
                dummy.next = ListNode(val1)
                l1 = l1.next 
            else: 
                dummy.next = ListNode(val2)
                l2 = l2.next 
        elif l1 != None: 
            val1 = l1.value 
            dummy.next = ListNode(val1)
            l1 = l1.next 
        elif l2 != None: 
            val2 = l2.value 
            dummy.next = ListNode(val2)
            l2 = l2.next 
        dummy = dummy.next 
    return x.next